# Setup

In [1]:
# Mount Google Drive and Change Working Directory to 'homework' Folder

from google.colab import drive
import sys

# Mount your Google Drive
drive.mount('/content/drive')

# # Define the root path to your "DATA4610" folder
root_path = '/content/drive/MyDrive/DATA4610/'

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
%%capture
# Install PostgreSQL
!apt-get -qq update > /dev/null
!apt-get -yq install postgresql > /dev/null

# Start the PostgreSQL service
!service postgresql start > /dev/null

# Install ipython-sql and psycopg2 for the %sql magic command
!pip install ipython-sql psycopg2-binary > /dev/null


# Create a user and database for our class
!sudo -u postgres psql -c "CREATE USER class_user WITH SUPERUSER PASSWORD 'password';"
!sudo -u postgres psql -c "CREATE DATABASE sample_db OWNER class_user;"

# Setup a database with name `sample_db` to be used
!sudo -u postgres psql -U postgres -c 'DROP DATABASE IF EXISTS sample_db;'
!sudo -u postgres psql -U postgres -c 'CREATE DATABASE sample_db;'

In [3]:

# Load the SQL magic extension
%load_ext sql

%config SqlMagic.style = '_DEPRECATED_DEFAULT'

# Define the connection string
connection_string = "postgresql://class_user:password@localhost/sample_db"

In [4]:
# Connect to the database using the magic command
%sql $connection_string

In [5]:
import pandas as pd
from sqlalchemy import create_engine, Integer, String, Date, Numeric, Text

# Assuming 'engine = create_engine(connection_string)' is defined right before this block runs.
engine = create_engine(connection_string)

# --- 1. DEFINE SQLALCHEMY DATA TYPES ---
# Explicitly map Pandas columns to precise SQL types to ensure correct schema creation
# (e.g., proper DATE types for date columns, NUMERIC for financial data).

SQL_DTYPES = {
    'flight_data_2024': {
        'year': Integer,
        'month': Integer,
        'day_of_month': Integer,
        'day_of_week': Integer,
        'fl_date': Date,
        'op_unique_carrier': String(10),
        'op_carrier_fl_num': Integer,
        'origin': String(10),
        'origin_city_name': String(100),
        'origin_state_nm': String(50),
        'dest': String(10),
        'dest_city_name': String(100),
        'dest_state_nm': String(50),
        'crs_dep_time': Integer,
        'dep_time': Numeric(10, 2),
        'dep_delay': Numeric(10, 2),
        'taxi_out': Numeric(10, 2),
        'wheels_off': Numeric(10, 2),
        'wheels_on': Numeric(10, 2),
        'taxi_in': Numeric(10, 2),
        'crs_arr_time': Integer,
        'arr_time': Numeric(10, 2),
        'arr_delay': Numeric(10, 2),
        'cancelled': Integer,
        'cancellation_code': String(10),
        'diverted': Integer,
        'crs_elapsed_time': Numeric(10, 2),
        'actual_elapsed_time': Numeric(10, 2),
        'air_time': Numeric(10, 2),
        'distance': Numeric(10, 2),
        'carrier_delay': Integer,
        'weather_delay': Integer,
        'nas_delay': Integer,
        'security_delay': Integer,
        'late_aircraft_delay': Integer
    }
}

# --- 2. LOAD CSVS AND PRE-PROCESS DATATYPES IN PANDAS (CHUNKED) ---
chunk_size = 100000  # Load 100k rows at a time

# Initialize an empty list to track success
chunks_loaded = 0

for chunk in pd.read_csv('/content/drive/MyDrive/DATA4610/data/flight_data_2024.csv',
                         chunksize=chunk_size):
    # Convert fl_date to datetime
    chunk['fl_date'] = pd.to_datetime(chunk['fl_date'], errors='coerce')

    # Convert numeric columns
    numeric_cols = ['dep_time', 'dep_delay', 'taxi_out', 'wheels_off', 'wheels_on',
                    'taxi_in', 'arr_time', 'arr_delay', 'crs_elapsed_time',
                    'actual_elapsed_time', 'air_time', 'distance']
    for col in numeric_cols:
        if col in chunk.columns:
            chunk[col] = pd.to_numeric(chunk[col], errors='coerce')

    # Append to SQL (use 'append' after first chunk, 'replace' for first)
    chunk.to_sql(
        'flight_data_2024',
        engine,
        if_exists='replace' if chunks_loaded == 0 else 'append',
        index=False,
        dtype=SQL_DTYPES['flight_data_2024']
    )

    chunks_loaded += 1
    print(f"Loaded chunk {chunks_loaded} ({chunk_size * chunks_loaded} total rows)")

print(f"✓ All {chunks_loaded} chunks loaded successfully!")

Loaded chunk 1 (100000 total rows)
Loaded chunk 2 (200000 total rows)
Loaded chunk 3 (300000 total rows)
Loaded chunk 4 (400000 total rows)
Loaded chunk 5 (500000 total rows)
Loaded chunk 6 (600000 total rows)


/tmp/ipython-input-1730512463.py:57: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('/content/drive/MyDrive/DATA4610/data/flight_data_2024.csv',


Loaded chunk 7 (700000 total rows)


/tmp/ipython-input-1730512463.py:57: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('/content/drive/MyDrive/DATA4610/data/flight_data_2024.csv',


Loaded chunk 8 (800000 total rows)


/tmp/ipython-input-1730512463.py:57: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('/content/drive/MyDrive/DATA4610/data/flight_data_2024.csv',


Loaded chunk 9 (900000 total rows)


/tmp/ipython-input-1730512463.py:57: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('/content/drive/MyDrive/DATA4610/data/flight_data_2024.csv',


Loaded chunk 10 (1000000 total rows)
Loaded chunk 11 (1100000 total rows)
Loaded chunk 12 (1200000 total rows)
Loaded chunk 13 (1300000 total rows)
Loaded chunk 14 (1400000 total rows)
Loaded chunk 15 (1500000 total rows)
Loaded chunk 16 (1600000 total rows)
Loaded chunk 17 (1700000 total rows)
Loaded chunk 18 (1800000 total rows)
Loaded chunk 19 (1900000 total rows)


/tmp/ipython-input-1730512463.py:57: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('/content/drive/MyDrive/DATA4610/data/flight_data_2024.csv',


Loaded chunk 20 (2000000 total rows)
Loaded chunk 21 (2100000 total rows)
Loaded chunk 22 (2200000 total rows)
Loaded chunk 23 (2300000 total rows)
Loaded chunk 24 (2400000 total rows)


/tmp/ipython-input-1730512463.py:57: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('/content/drive/MyDrive/DATA4610/data/flight_data_2024.csv',


Loaded chunk 25 (2500000 total rows)
Loaded chunk 26 (2600000 total rows)


/tmp/ipython-input-1730512463.py:57: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('/content/drive/MyDrive/DATA4610/data/flight_data_2024.csv',


Loaded chunk 27 (2700000 total rows)
Loaded chunk 28 (2800000 total rows)
Loaded chunk 29 (2900000 total rows)
Loaded chunk 30 (3000000 total rows)
Loaded chunk 31 (3100000 total rows)
Loaded chunk 32 (3200000 total rows)
Loaded chunk 33 (3300000 total rows)
Loaded chunk 34 (3400000 total rows)
Loaded chunk 35 (3500000 total rows)
Loaded chunk 36 (3600000 total rows)
Loaded chunk 37 (3700000 total rows)
Loaded chunk 38 (3800000 total rows)
Loaded chunk 39 (3900000 total rows)


/tmp/ipython-input-1730512463.py:57: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('/content/drive/MyDrive/DATA4610/data/flight_data_2024.csv',


Loaded chunk 40 (4000000 total rows)
Loaded chunk 41 (4100000 total rows)
Loaded chunk 42 (4200000 total rows)
Loaded chunk 43 (4300000 total rows)
Loaded chunk 44 (4400000 total rows)
Loaded chunk 45 (4500000 total rows)
Loaded chunk 46 (4600000 total rows)
Loaded chunk 47 (4700000 total rows)
Loaded chunk 48 (4800000 total rows)
Loaded chunk 49 (4900000 total rows)
Loaded chunk 50 (5000000 total rows)


/tmp/ipython-input-1730512463.py:57: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('/content/drive/MyDrive/DATA4610/data/flight_data_2024.csv',


Loaded chunk 51 (5100000 total rows)
Loaded chunk 52 (5200000 total rows)
Loaded chunk 53 (5300000 total rows)
Loaded chunk 54 (5400000 total rows)
Loaded chunk 55 (5500000 total rows)


/tmp/ipython-input-1730512463.py:57: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('/content/drive/MyDrive/DATA4610/data/flight_data_2024.csv',


Loaded chunk 56 (5600000 total rows)


/tmp/ipython-input-1730512463.py:57: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('/content/drive/MyDrive/DATA4610/data/flight_data_2024.csv',


Loaded chunk 57 (5700000 total rows)
Loaded chunk 58 (5800000 total rows)
Loaded chunk 59 (5900000 total rows)
Loaded chunk 60 (6000000 total rows)


/tmp/ipython-input-1730512463.py:57: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('/content/drive/MyDrive/DATA4610/data/flight_data_2024.csv',


Loaded chunk 61 (6100000 total rows)


/tmp/ipython-input-1730512463.py:57: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('/content/drive/MyDrive/DATA4610/data/flight_data_2024.csv',


Loaded chunk 62 (6200000 total rows)


/tmp/ipython-input-1730512463.py:57: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('/content/drive/MyDrive/DATA4610/data/flight_data_2024.csv',


Loaded chunk 63 (6300000 total rows)


/tmp/ipython-input-1730512463.py:57: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('/content/drive/MyDrive/DATA4610/data/flight_data_2024.csv',


Loaded chunk 64 (6400000 total rows)
Loaded chunk 65 (6500000 total rows)
Loaded chunk 66 (6600000 total rows)
Loaded chunk 67 (6700000 total rows)


/tmp/ipython-input-1730512463.py:57: DtypeWarning: Columns (24) have mixed types. Specify dtype option on import or set low_memory=False.
  for chunk in pd.read_csv('/content/drive/MyDrive/DATA4610/data/flight_data_2024.csv',


Loaded chunk 68 (6800000 total rows)
Loaded chunk 69 (6900000 total rows)
Loaded chunk 70 (7000000 total rows)
Loaded chunk 71 (7100000 total rows)
✓ All 71 chunks loaded successfully!


# Flight Delay Patterns Analysis
___

### About the Dataset

This dataset contains detailed flight performance and delay information for domestic flights in 2024, merged from monthly BTS TranStats files into a single cleaned dataset. It includes over 7 million rows and 35 columns, providing comprehensive information on scheduled and actual flight times, delays, cancellations, diversions, and distances between airports. The dataset is suitable for exploratory data analysis (EDA), machine learning tasks such as delay prediction, time series analysis, and airline/airport performance studies.

---

### The Questions I'm Answering
1. How do delays build up throughout the day? Analyze whether late aircraft delays increase as the day goes on by tracking cumulative delay patterns by carrier and route, and examine if morning delays lead to more delays in the afternoon and evening.
2. How do delay patterns differ between hub-to-hub, hub-to-regional, and regional-to-regional routes?
3. How do flight delays vary on major holidays compared to regular days?
4. Which day of the week experiences the largest flight delays?


## Question 1

In [9]:
%%sql
-- Analyze how delays build up throughout the day (no carrier/route breakdown)
WITH time_of_day_delays AS (
    SELECT
        -- Categorize by time of day
        CASE
            WHEN crs_dep_time BETWEEN 500 AND 1159 THEN 'Morning'
            WHEN crs_dep_time BETWEEN 1200 AND 1659 THEN 'Afternoon'
            WHEN crs_dep_time BETWEEN 1700 AND 2100 THEN 'Evening'
            ELSE 'Night/Early Morning'
        END AS time_period,
        dep_delay,
        arr_delay,
        late_aircraft_delay,
        carrier_delay,
        weather_delay,
        nas_delay
    FROM flight_data_2024
    WHERE cancelled = 0  -- not including cancelled flights
        AND dep_delay IS NOT NULL
        AND arr_delay IS NOT NULL
)

SELECT
    time_period,
    COUNT(*) AS num_flights,
    ROUND(AVG(dep_delay), 2) AS avg_dep_delay,
    ROUND(AVG(arr_delay), 2) AS avg_arr_delay,
    ROUND(AVG(late_aircraft_delay), 2) AS avg_late_aircraft_delay,
    ROUND(AVG(carrier_delay), 2) AS avg_carrier_delay,
    ROUND(AVG(weather_delay), 2) AS avg_weather_delay,
    ROUND(AVG(nas_delay), 2) AS avg_nas_delay
FROM time_of_day_delays
GROUP BY time_period
ORDER BY
    CASE time_period
        WHEN 'Morning' THEN 1
        WHEN 'Afternoon' THEN 2
        WHEN 'Evening' THEN 3
        ELSE 4
    END;

 * postgresql://class_user:***@localhost/sample_db
4 rows affected.


[('Morning', 2907959, Decimal('6.48'), Decimal('0.60'), Decimal('2.73'), Decimal('4.43'), Decimal('0.62'), Decimal('2.07')),
 ('Afternoon', 2033975, Decimal('14.31'), Decimal('9.28'), Decimal('7.00'), Decimal('4.92'), Decimal('0.94'), Decimal('3.36')),
 ('Evening', 1570965, Decimal('19.73'), Decimal('14.77'), Decimal('10.13'), Decimal('6.06'), Decimal('1.24'), Decimal('3.67')),
 ('Night/Early Morning', 452368, Decimal('19.32'), Decimal('12.44'), Decimal('8.59'), Decimal('7.58'), Decimal('1.17'), Decimal('2.23'))]

### Analysis


The analysis reveals a clear cascade effect where delays build up throughout the day. Morning flights (5 AM - noon) start with an average late aircraft delay of just 2.73 minutes, but this more than triples to 10.13 minutes by evening (5 PM - 9 PM). This pattern confirms that when aircraft are delayed on earlier flights, those delays ripple forward to later departures. Evening flights experience the worst delays overall, with average departure delays of 19.73 minutes and arrival delays of 14.77 minutes. The late aircraft delay metric, which specifically tracks delays caused by the previous flight arriving late, is the key evidence showing how morning disruptions compound into significant delays by the end of the day.

##  Question 2


In [12]:
%%sql
-- Step 1: Create a list of all US hub airports
WITH hub_airports AS (
    SELECT airport_code FROM (VALUES
        -- United Airlines
        ('DEN'), ('IAH'), ('LAX'), ('EWR'), ('ORD'), ('SFO'), ('IAD'),
        -- American Airlines
        ('CLT'), ('DFW'), ('MIA'), ('JFK'), ('LGA'), ('PHL'), ('PHX'), ('DCA'),
        -- Delta
        ('CVG'), ('DTW'), ('ATL'), ('BOS'), ('MSP'), ('SLC'), ('SEA'),
        -- Southwest
        ('DAL'), ('MDW'), ('HOU'), ('BWI'), ('LAS'), ('OAK'), ('MCO'),
        -- JetBlue
        ('FLL'), ('LGB'), ('SJU'),
        -- Alaska Airlines
        ('PDX'), ('ANC'), ('SAN'), ('SJC')
    ) AS hubs(airport_code)
),

-- Step 2: Classify each route by type
route_classification AS (
    SELECT
        f.*,
        CASE
            WHEN o.airport_code IS NOT NULL AND d.airport_code IS NOT NULL
                THEN 'Hub-to-Hub'
            WHEN o.airport_code IS NOT NULL OR d.airport_code IS NOT NULL
                THEN 'Hub-to-Regional'
            ELSE 'Regional-to-Regional'
        END AS route_type
    FROM flight_data_2024 f
    LEFT JOIN hub_airports o ON f.origin = o.airport_code
    LEFT JOIN hub_airports d ON f.dest = d.airport_code
    WHERE f.cancelled = 0
        AND f.dep_delay IS NOT NULL
        AND f.arr_delay IS NOT NULL
)

-- Step 3: Calculate delay metrics by route type
SELECT
    route_type,
    COUNT(*) AS num_flights,
    ROUND(AVG(dep_delay), 2) AS avg_dep_delay,
    ROUND(AVG(arr_delay), 2) AS avg_arr_delay,
    ROUND(AVG(late_aircraft_delay), 2) AS avg_late_aircraft_delay,
    ROUND(AVG(carrier_delay), 2) AS avg_carrier_delay,
    ROUND(AVG(weather_delay), 2) AS avg_weather_delay,
    ROUND(AVG(nas_delay), 2) AS avg_nas_delay,
    -- Calculate on-time percentage (arrival delay <= 15 minutes)
    ROUND(100.0 * SUM(CASE WHEN arr_delay <= 15 THEN 1 ELSE 0 END) / COUNT(*), 2) AS on_time_pct
FROM route_classification
GROUP BY route_type
ORDER BY
    CASE route_type
        WHEN 'Hub-to-Hub' THEN 1
        WHEN 'Hub-to-Regional' THEN 2
        ELSE 3
    END;

 * postgresql://class_user:***@localhost/sample_db
3 rows affected.


[('Hub-to-Hub', 2656928, Decimal('14.04'), Decimal('7.85'), Decimal('6.30'), Decimal('5.22'), Decimal('0.69'), Decimal('3.48'), Decimal('78.32')),
 ('Hub-to-Regional', 3943748, Decimal('11.93'), Decimal('6.78'), Decimal('5.89'), Decimal('5.23'), Decimal('1.04'), Decimal('2.47'), Decimal('80.61')),
 ('Regional-to-Regional', 364591, Decimal('9.14'), Decimal('5.05'), Decimal('5.54'), Decimal('3.61'), Decimal('0.73'), Decimal('1.76'), Decimal('82.76'))]

### Analysis

The analysis reveals that hub-to-hub routes experience the most significant delays, while regional-to-regional routes perform best. Hub-to-hub flights averaged 14.04 minutes departure delay and 7.85 minutes arrival delay, with only 78.32% arriving on-time (within 15 minutes of schedule). This higher delay rate is likely due to congestion at major airports and the complexity of coordinating connecting flights. Hub-to-regional routes showed moderate performance with 11.93 minute average departure delays and 80.61% on-time arrivals. Surprisingly, regional-to-regional routes had the best performance with only 9.14 minute average departure delays, 5.05 minute arrival delays, and the highest on-time percentage at 82.76%. Interestingly, carrier delays and late aircraft delays remain fairly consistent across all route types (around 5-6 minutes), while NAS delays are notably higher for hub-to-hub routes (3.48 min vs 1.76 min for regional), suggesting that air traffic congestion at major hubs is a key contributing factor to delays.

## Question 3

In [14]:
%%sql
-- Step 1: Define major US holidays in 2024
WITH holidays_2024 AS (
    SELECT holiday_name, holiday_date FROM (VALUES
        ('New Years', DATE '2024-01-01'),
        ('MLK Day', DATE '2024-01-15'),
        ('Memorial Day', DATE '2024-05-27'),
        ('July 4th', DATE '2024-07-04'),
        ('Labor Day', DATE '2024-09-02'),
        ('Thanksgiving', DATE '2024-11-28'),
        ('Christmas', DATE '2024-12-25')
    ) AS h(holiday_name, holiday_date)
),

-- Step 2: Create holiday windows (3 days before and 2 days after)
holiday_windows AS (
    SELECT
        holiday_name,
        holiday_date,
        holiday_date - INTERVAL '3 days' AS window_start,
        holiday_date + INTERVAL '2 days' AS window_end
    FROM holidays_2024
),

-- Step 3: Classify flights as holiday or regular
flight_classification AS (
    SELECT
        f.fl_date,
        f.dep_delay,
        f.arr_delay,
        f.cancelled,
        f.late_aircraft_delay,
        f.carrier_delay,
        f.weather_delay,
        f.nas_delay,
        CASE
            WHEN EXISTS (
                SELECT 1 FROM holiday_windows hw
                WHERE f.fl_date BETWEEN hw.window_start AND hw.window_end
            ) THEN 'Holiday Period'
            ELSE 'Regular Day'
        END AS day_type
    FROM flight_data_2024 f
    WHERE f.dep_delay IS NOT NULL
        AND f.arr_delay IS NOT NULL
)

-- Step 4: Compare holiday vs regular day delays
SELECT
    day_type,
    COUNT(*) AS num_flights,
    ROUND(AVG(dep_delay), 2) AS avg_dep_delay,
    ROUND(AVG(arr_delay), 2) AS avg_arr_delay,
    ROUND(AVG(late_aircraft_delay), 2) AS avg_late_aircraft_delay,
    ROUND(AVG(carrier_delay), 2) AS avg_carrier_delay,
    ROUND(AVG(weather_delay), 2) AS avg_weather_delay,
    ROUND(AVG(nas_delay), 2) AS avg_nas_delay,
    -- Cancellation rate
    ROUND(100.0 * SUM(cancelled) / (COUNT(*) + SUM(cancelled)), 2) AS cancellation_rate,
    -- On-time percentage
    ROUND(100.0 * SUM(CASE WHEN arr_delay <= 15 THEN 1 ELSE 0 END) / COUNT(*), 2) AS on_time_pct
FROM flight_classification
GROUP BY day_type
ORDER BY day_type DESC;

 * postgresql://class_user:***@localhost/sample_db
2 rows affected.


[('Regular Day', 6255562, Decimal('12.29'), Decimal('6.78'), Decimal('5.91'), Decimal('5.07'), Decimal('0.81'), Decimal('2.78'), Decimal('0.00'), Decimal('80.08')),
 ('Holiday Period', 709705, Decimal('15.27'), Decimal('9.90'), Decimal('7.01'), Decimal('5.80'), Decimal('1.61'), Decimal('3.14'), Decimal('0.00'), Decimal('77.83'))]

In [13]:
%%sql
-- Detailed breakdown by individual holiday
WITH holidays_2024 AS (
    SELECT holiday_name, holiday_date FROM (VALUES
        ('New Years', DATE '2024-01-01'),
        ('MLK Day', DATE '2024-01-15'),
        ('Memorial Day', DATE '2024-05-27'),
        ('July 4th', DATE '2024-07-04'),
        ('Labor Day', DATE '2024-09-02'),
        ('Thanksgiving', DATE '2024-11-28'),
        ('Christmas', DATE '2024-12-25')
    ) AS h(holiday_name, holiday_date)
),

holiday_windows AS (
    SELECT
        holiday_name,
        holiday_date,
        holiday_date - INTERVAL '3 days' AS window_start,
        holiday_date + INTERVAL '2 days' AS window_end
    FROM holidays_2024
)

SELECT
    hw.holiday_name,
    COUNT(*) AS num_flights,
    ROUND(AVG(f.dep_delay), 2) AS avg_dep_delay,
    ROUND(AVG(f.arr_delay), 2) AS avg_arr_delay,
    ROUND(AVG(f.late_aircraft_delay), 2) AS avg_late_aircraft_delay,
    ROUND(100.0 * SUM(CASE WHEN f.arr_delay <= 15 THEN 1 ELSE 0 END) / COUNT(*), 2) AS on_time_pct
FROM flight_data_2024 f
JOIN holiday_windows hw ON f.fl_date BETWEEN hw.window_start AND hw.window_end
WHERE f.cancelled = 0
    AND f.dep_delay IS NOT NULL
    AND f.arr_delay IS NOT NULL
GROUP BY hw.holiday_name
ORDER BY avg_arr_delay DESC;

 * postgresql://class_user:***@localhost/sample_db
7 rows affected.


[('MLK Day', 92936, Decimal('33.09'), Decimal('31.62'), Decimal('13.92'), Decimal('60.67')),
 ('Memorial Day', 113067, Decimal('20.65'), Decimal('17.53'), Decimal('9.70'), Decimal('72.28')),
 ('Christmas', 111059, Decimal('17.26'), Decimal('9.72'), Decimal('7.81'), Decimal('77.84')),
 ('July 4th', 114692, Decimal('11.01'), Decimal('4.65'), Decimal('5.43'), Decimal('81.88')),
 ('Labor Day', 110604, Decimal('9.68'), Decimal('4.09'), Decimal('4.84'), Decimal('82.97')),
 ('Thanksgiving', 112700, Decimal('6.67'), Decimal('0.38'), Decimal('3.25'), Decimal('85.12')),
 ('New Years', 54647, Decimal('7.75'), Decimal('-0.02'), Decimal('3.57'), Decimal('84.59'))]

### Analysis

Holiday periods experience noticeably worse delays compared to regular days, with average departure delays of 15.27 minutes versus 12.29 minutes, arrival delays of 9.90 minutes versus 6.78 minutes, and on-time performance dropping from 80.08% to 77.83%. Late aircraft delays increase by nearly 20% during holidays (7.01 min vs 5.91 min), weather delays more than double (1.61 min vs 0.81 min), and NAS delays increase by 13% (3.14 min vs 2.78 min), suggesting that higher passenger volumes create airport congestion while making it harder to recover from disruptions. However, the impact varies dramatically by holiday: MLK Day is by far the worst with severe delays averaging 33.09 minutes for departures and 31.62 minutes for arrivals (only 60.67% on-time), likely due to winter weather conditions in mid-January. Memorial Day and Christmas also show significant disruptions with 17-20 minute departure delays, while July 4th and Labor Day experience moderate impacts. Surprisingly, Thanksgiving and New Years perform exceptionally well. Thanksgiving actually outperforms regular days with only 6.67 minute departure delays and 85.12% on-time performance, and New Years shows minimal impact with 7.75 minute departure delays and 84.59% on-time performance. This suggests that winter weather, rather than passenger volume alone, is the primary driver of holiday-related delays.

## Question 4

In [16]:
%%sql
WITH daily_stats AS (
    SELECT
        day_of_week,
        CASE day_of_week
            WHEN 1 THEN 'Monday'
            WHEN 2 THEN 'Tuesday'
            WHEN 3 THEN 'Wednesday'
            WHEN 4 THEN 'Thursday'
            WHEN 5 THEN 'Friday'
            WHEN 6 THEN 'Saturday'
            WHEN 7 THEN 'Sunday'
        END AS day_name,
        COUNT(*) AS num_flights,
        ROUND(AVG(dep_delay), 2) AS avg_dep_delay,
        ROUND(AVG(arr_delay), 2) AS avg_arr_delay,
        ROUND(AVG(late_aircraft_delay), 2) AS avg_late_aircraft_delay
    FROM flight_data_2024
    WHERE cancelled = 0
        AND dep_delay IS NOT NULL
        AND arr_delay IS NOT NULL
    GROUP BY day_of_week
)

SELECT
    day_name,
    num_flights,
    avg_dep_delay,
    avg_arr_delay,
    avg_late_aircraft_delay,
    ROUND(AVG(avg_dep_delay) OVER (), 2) AS weekly_avg_dep_delay,
    ROUND(avg_dep_delay - AVG(avg_dep_delay) OVER (), 2) AS diff_from_weekly_avg,
    RANK() OVER (ORDER BY avg_arr_delay DESC) AS delay_rank
FROM daily_stats
ORDER BY day_of_week;

 * postgresql://class_user:***@localhost/sample_db
7 rows affected.


[('Monday', 1056398, Decimal('12.64'), Decimal('7.03'), Decimal('6.04'), Decimal('12.54'), Decimal('0.10'), 4),
 ('Tuesday', 961064, Decimal('10.36'), Decimal('4.39'), Decimal('5.15'), Decimal('12.54'), Decimal('-2.18'), 7),
 ('Wednesday', 958699, Decimal('10.33'), Decimal('4.83'), Decimal('5.01'), Decimal('12.54'), Decimal('-2.21'), 6),
 ('Thursday', 1024394, Decimal('13.08'), Decimal('8.18'), Decimal('6.37'), Decimal('12.54'), Decimal('0.54'), 3),
 ('Friday', 1037087, Decimal('14.59'), Decimal('9.61'), Decimal('6.99'), Decimal('12.54'), Decimal('2.05'), 1),
 ('Saturday', 898255, Decimal('12.51'), Decimal('6.45'), Decimal('5.77'), Decimal('12.54'), Decimal('-0.03'), 5),
 ('Sunday', 1029370, Decimal('14.28'), Decimal('8.77'), Decimal('6.68'), Decimal('12.54'), Decimal('1.74'), 2)]

### Analysis

Friday experiences the largest flight delays with average departure delays of 14.59 minutes and arrival delays of 9.61 minutes, ranking as the worst day of the week (2.05 minutes above the weekly average). Sunday follows closely as the second-worst day with 14.28 minute departure delays and 8.77 minute arrival delays. Tuesday and Wednesday are the best-performing days with average departure delays around 10.3 minutes, more than 2 minutes below the weekly average and the lowest arrival delays at 4.39 and 4.83 minutes respectively. The pattern reveals that weekends and Fridays are significantly worse than mid-week travel, with late aircraft delays peaking on Friday (6.99 min) and Sunday (6.68 min), suggesting that the combination of higher leisure travel volume and accumulated delays from the week compound to create the worst performance. Monday and Thursday fall in the middle range, while the Tuesday-Wednesday window offers travelers the most reliable on-time performance.